pkg/reconciler/pubsub/topic/topic.go

```diff
@@ -0,0 +1,441 @@
+			"Failed to update finalizers for Topic %q: %v", topic.Name, fErr)
+		return fErr
+	} else if updated {
+		// There was a difference and updateFinalizers said it updated and did not return an error.
+		r.Recorder.Eventf(topic, corev1.EventTypeNormal, "Updated", "Updated Topic %q finalizers", topic.Name)
+	}
+
+	if equality.Semantic.DeepEqual(original.Status, topic.Status) {
+		// If we didn't change anything then don't call updateStatus.
+		// This is important because the copy we loaded from the informer's
+		// cache may be stale and we don't want to overwrite a prior update
+		// to status with this stale state.
+
+	} else if _, uErr := r.updateStatus(ctx, topic); uErr != nil {
+		logging.FromContext(ctx).Desugar().Warn("Failed to update Topic status", zap.Error(uErr))
+		r.Recorder.Eventf(topic, corev1.EventTypeWarning, "UpdateFailed",
+			"Failed to update status for Topic %q: %v", topic.Name, uErr)
+		return uErr
+	} else if reconcileErr == nil {
+		// There was a difference and updateStatus did not return an error.
+		r.Recorder.Eventf(topic, corev1.EventTypeNormal, "Updated", "Updated Topic %q", topic.Name)
+	}
+	if reconcileErr != nil {
+		r.Recorder.Event(topic, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
+	}
+	return reconcileErr
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, topic *v1alpha1.Topic) error {
+	ctx = logging.WithLogger(ctx, r.Logger.With(zap.Any("topic", topic)))
+
+	topic.Status.InitializeConditions()
+
+	if topic.DeletionTimestamp != nil {
+		logging.FromContext(ctx).Desugar().Debug("Deleting Pub/Sub topic")
+		if topic.Spec.PropagationPolicy == v1alpha1.TopicPolicyCreateDelete {
+			if err := r.deleteTopic(ctx, topic); err != nil {
+				topic.Status.MarkNoTopic("TopicDeleteFailed", "Failed to delete Pub/Sub topic: %s", err.Error())
+				return err
+			}
+			topic.Status.MarkNoTopic("TopicDeleted", "Successfully deleted Pub/Sub topic: %s", topic.Status.TopicID)
+			topic.Status.TopicID = ""
+		}
+		removeFinalizer(topic)
+		return nil
+	}
+
+	// Add the finalizer.
+	addFinalizer(topic)
+
+	if err := r.reconcileTopic(ctx, topic); err != nil {
+		topic.Status.MarkNoTopic("TopicCreateFailed", "Failed to create Pub/Sub topic: %s", err.Error())
+		return err
+	}
+	topic.Status.MarkTopicReady()
+	// Set the topic being used.
+	topic.Status.TopicID = topic.Spec.Topic
+
+	err, svc := r.reconcilePublisher(ctx, topic)
+	if err != nil {
+		topic.Status.MarkNotDeployed("PublisherCreateFailed", "Failed to create Publisher: %s", err.Error())
+		return err
+	}
+
+	// Update the topic.
+	topic.Status.PropagatePublisherStatus(svc.Status.GetCondition(apis.ConditionReady))
+	if svc.Status.IsReady() {
+		topic.Status.SetAddress(svc.Status.Address.URL)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) reconcileTopic(ctx context.Context, topic *v1alpha1.Topic) error {
+	if topic.Status.ProjectID == "" {
+		projectID, err := utils.ProjectID(topic.Spec.Project)
```

Comment:
Assuming:
- The project id cannot change over the lifetime of the cluster
- The webhook's project id is always the same as the controller's project id

then I suggest doing this defaulting in the webhook instead of the reconcilers.

Advantages:
- All defaulting is in one place (either in the per-kind APIs or shared across all kinds)
- If the project ID can't be defaulted (running outside GCE), the user sees that right away as a create error
- Fewer queries to the metadata server and decreased reliance on its availability. If metadata server goes down, existing resources can still be reconciled.

Disadvantages:
- Eliminates the ability to dump the resources from the cluster as yaml and apply them to a cluster in a different project without deleting the ProjectID field. This seems ok since (AFAIK) this is not generally possible in K8s anyway (metadata fields like uuid need to be removed)

This suggestion applies to all other reconcilers that have a similar defaulting.

Reply:
This is another valid point and you are right, it will make things cleaner. I also like the fail fast approach.

I'm not entirely sure if it's valid the assumption that the project id given by the metadata server will always be the same throughout the lifetime of the cluster, thus we may be adding state to the webhook. I recall we had some of these conversations in the past, but don't actually remember the context then. Do you?

Even if the project id can change, do you still think it's better to have it on the webhook? 
Let me ask Scott why they didn't default it in the webhook, just in case we are missing something.

Reply:
If the project id can change, then I think we should leave the current reconciler defaulting. I'd be surprised if that's possible though.

Either way, this can be a followup. I created #450.

Reply:
Hmm, isn't the project ID returned by the metadata server the default one? If you have multiple projects and you change the default, then wouldn't that return something different?
I might be wrong though

Reply:
What do you mean by "change the default [project]"? Are you referring to a GCP setting?

Reply:
Let me take that back. I think you are right, and that cannot be changed. 
I was just talking with @yolocs and looked at https://cloud.google.com/compute/docs/storing-retrieving-metadata...
`Compute Engine defines a set of default metadata entries that provide information about your instance or project. Default metadata is always defined and set by the server. You can't manually edit any of these metadata pairs.`

Reply:
Resolving as it will be done in a follow up